### DISCUSSION TITLE
[Question] Is there any reasoning behind using `"display": "minimal-ui"` in the manifest?

### DISCUSSION THREAD

**Discussion by aivarovskiy** [schlagmichdoch/PairDrop#273] [Ideas]
_I feel like I've seen similar issues but couldn't find them so I opened a new one._

**Problem**
If I install PairDrop as a PWA, because of `"display": "minimal-ui"` in [manifest.json](https://github.com/schlagmichdoch/PairDrop/blob/3fa0873bc4f71f306f0f72afdfecc283bff75b85/public/manifest.json#L31) it opens a new tab in Safari instead of opening its own window. Because of this behavior, there is no point in using PairDrop as a PWA (at least for me).

**Possible solution**
Other PWA apps I've installed use `"display": "standalone"` which solves this issue.

**Question**
Is there any particular reason for using `"minimal-ui"` instead of `"standalone"`?

**Smartphone**
- Device: iPhone X
- OS: IOS 16.1.2
- Browser: Safari

https://github.com/schlagmichdoch/PairDrop/blob/3fa0873bc4f71f306f0f72afdfecc283bff75b85/public/manifest.json#L31

**Comment by schlagmichdoch** [comment]
`"display": "standalone"` puts the PWA into its own context so devices paired in the PWA are not the same as devices paired using PairDrop with Safari. If you e.g. pair devices using the QR code you cannot open the given link in a PWA.

I think this might be confusing behavior for users, especially if they want to use the Siri shortcut as well.

Installing the PWA is useful anyway, as this prevents Safari from purging the data every now and then which makes pairings and name changes persistent.

**Comment by aivarovskiy** [comment]
Thanks for the explanation!

But that's unfortunate. As for me I would rather get rid of QR codes completely. It's not like it's much easier to use camera than just typing in 5-6 digits. I feel like it's even harder to be honest.

Still very nice piece of software. Keep it up!

**Comment by schlagmichdoch** [comment]
I understand your point of view and I’m with you that most users will probably not use the QR-codes as of now.

There are some arguments to keep them:
- Possible future use cases for the QR-codes would not require changes to the UI
- QR-codes are a clear indicator that the section and the pair code shown is meant for another device.
- when you click/tap on the QR codes, a direkt link is copied to your clipboard.

In the future I might add further encryption (#180) and user validation to make PairDrop a zero trust application (as in you wouldn‘t need to trust the server for signaling).

You would then have the possibility to show a persistent security number for any connection of two devices, that you can compare and mark as verified (Analog to the Signal security number).

Alongside the pair key, the QR code will then include an additional symmetric key that is used to secure the key exchange. This will mark the paired device directly as verified and will reduce the user flow by one step if you will use the QR code.

There could also be the possibility to add a QR code scanner to PairDrop itself. This would make it possible to use QR codes from inside the PWA or an app (like Snapdrop & PairDrop for Android).

I’m not sure where to put a camera button though in order to not clutter the UI. Any thoughts?

**Comment by schlagmichdoch** [comment]
Also, do you host PairDrop yourself or do you use pairdrop.net? I guess we could make the manifest variable `"display": "standalone"` editable via an environment variable.

**Comment by aivarovskiy** [comment]
> There could also be the possibility to add a QR code scanner to PairDrop itself. This would make it possible to use QR codes from inside the PWA or an app (like Snapdrop & PairDrop for Android).

That occured to me too, but I thought it went against the idea of simplicity that you talk about in FAQ.

> I’m not sure where to put a camera button though in order to not clutter the UI. Any thoughts?

To be honest I don't know. Here are my thoughts:
- Adding another icon at the top feels wrong. I think there are already almost too many icons.
- Making the camera accessible only from pairing/room dialogs adds extra step and makes it more confusing to find.
- Tab bar with a separate tab for the camera sounds like a good option, but that, again, goes against the idea of simplicity. However, in return it'll help clean up the UI and improve the UX.

> Also, do you host PairDrop yourself or do you use pairdrop.net? I guess we could make the manifest variable "display": "standalone" editable via an environment variable.

I use pairdrop.net. I guess that's not a bad idea, especially if there's an explanation of what it will change.

**Comment by schlagmichdoch** [comment]
Thanks for your thoughts!

> That occured to me too, but I thought it went against the idea of simplicity that you talk about in FAQ.

Yes true, it’s not easy to find the right balance and keep the main user flow straight forward.

In the end, an in-app QR-code scanner would only be an improvement for the PWA and for (future) apps. So it could be a compromise to show the camera button only when these states are recognized. (We could easily check that the PWA is running standalone and an app could use an API call.)

So it would never be shown to users that use the browser or on Desktop browsers (which would not make much sense) but only from the app.

> * Adding another icon at the top feels wrong. I think there are already almost too many icons.
> * Making the camera accessible only from pairing/room dialogs adds extra step and makes it more confusing to find.
> * Tab bar with a separate tab for the camera sounds like a good option, but that, again, goes against the idea of simplicity. However, in return it'll help clean up the UI and improve the UX.

Good points. A tab bar would be on the bottom of everything else, wouldn’t it? I’m resistent to use up even more vertical space as the most space should be reserved for the peers..

My current favorite would be on the top left corner analogous to the about PairDrop (x) button. Any thoughts?



> I guess that's not a bad idea, especially if there's an explanation of what it will change.

Sure there would be a reference in the docs

**Comment by aivarovskiy** [comment]
> So it would never be shown to users that use the browser or on Desktop browsers (which would not make much sense) but only from the app.

Didn't quite get it. Why wouldn't it make sense? Because of tablets and laptops with rear cameras?

**Comment by aivarovskiy** [comment]
> Good points. A tab bar would be on the bottom of everything else, wouldn’t it? I’m resistent to use up even more vertical space as the most space should be reserved for the peers..

Yeah, on the bottom like in IOS apps. Maybe you're right.

> My current favorite would be on the top left corner analogous to the about PairDrop (x) button. Any thoughts?

I guess that's the best solution if you don't feel like adding a tab bar.

**Comment by schlagmichdoch** [comment]
> Didn't quite get it. Why wouldn't it make sense? Because of tablets and laptops with rear cameras?

I don’t think anyone would scan a QR code via the front camera of a laptop so I would hide the button on Desktop. I’m not sure about Tablets. 

I guess the best approach would be to only show the button if display is standalone and device is Android / iOS.

It’s a small feature though that would add complexity. I will put this into ‚to be decided‘ for now.

**Comment by schlagmichdoch** [comment]
I have decided to add the standalone feature to the next minor release `v1.11.0`. Stay tuned!
